packages/plugin-ext/src/main/browser/menus/menus-contribution-handler.ts

```diff
@@ -108,146 +107,189 @@ export class MenusContributionPointHandler {
+                toDispose.push(
+                    this.registerMenusWithSubmenus(plugin, location, (menu, menuPath) =>
+                        this.registerMenuAction(menuPath!, menu, command => ({
                             execute: (...args) => this.commands.executeCommand(command, args[0]),
                             isEnabled: (...args) => this.commands.isEnabled(command, args[0]),
                             isVisible: (...args) => this.commands.isVisible(command, args[0])
-                        })));
-                    }
-                }
+                        })),
+                        [DebugStackFramesWidget.CONTEXT_MENU, DebugThreadsWidget.CONTEXT_MENU]
+                    )
+                );
             } else if (allMenus.hasOwnProperty(location)) {
                 const menuPaths = MenusContributionPointHandler.parseMenuPaths(location);
                 if (!menuPaths.length) {
                     this.logger.warn(`'${plugin.metadata.model.id}' plugin contributes items to a menu with invalid identifier: ${location}`);
                     continue;
                 }
-                const menus = allMenus[location];
-                menus.forEach(menu => {
-                    for (const menuPath of menuPaths) {
-                        toDispose.push(this.registerGlobalMenuAction(menuPath, menu));
-                    }
-                });
+                toDispose.push(
+                    this.registerMenusWithSubmenus(plugin, location, (menu, menuPath) =>
+                        this.registerGlobalMenuAction(menuPath!, menu),
+                        menuPaths)
+                );
             }
         }
         return toDispose;
     }
 
+    protected registerMenusWithSubmenus(plugin: DeployedPlugin, location: string, registerMenu: (menu: Menu, menuPath?: MenuPath) => Disposable,
+        menuPaths?: MenuPath[]): Disposable {
+
+        const allMenus = plugin.contributes && plugin.contributes.menus;
+        if (!allMenus) {
+            return Disposable.NULL;
+        }
+        const allSubmenus = plugin.contributes && plugin.contributes.submenus;
+        const toDispose = new DisposableCollection();
+        // Set of registered submenus ids, In order to discover duplicates/cycles.
+        const submenusIds = new Set<string>();
+
+        const registerActions = (menus: Menu[], group: string | undefined) => {
+            menus.forEach(menu => {
+                let menuPath; let menuGroup = menu.group; let index = 0;
+                do {
+                    if (menuPaths) {
+                        menuPath = menuPaths[index++];
+                    }
+                    if (group) {
+                        menuGroup = group + (menuGroup ? '/' + menuGroup.split('@')[0] : '/_');
+                    }
+                    if (menu.submenu) {
+                        if (submenusIds.has(menu.submenu)) {
```

Comment:
Do we want to strictly forbid all repetition? I could imagine a scenario (though it's a bit far fetched) where it might be desirable to do something like this:
```
└── menu
    ├── option
    ├── submenu1
    │   └── option1
    └── submenu2
        ├── option2
        └── submenu1 // In case you expected to find it here
            └── option1
```
Maybe only preventing cycles absolutely and repetition at the same level is important, but repetitions at different levels should be at the discretion of the plugin author?

Reply:
There won't be a problem in this case, only if `submenu1` has submenu `submenu2` the cycle will be prevented.

Reply:
Since the `submenuIds` set is used throughout the entire registration of all submenus for a particular contribution point, I believe only one instance of a given submenu can appear in any given submenu collection. For example, I created submenus with the following manifest:

```jsonc
{
        "submenus": [
            {
                "id": "ep.custosubmenu1",
                "label": "custom submenu 1"
            },
            {
                "id": "ep.custosubmenu2",
                "label": "custom submenu 2"
            },
            {
                "id": "cjg.customsubmenu1",
                "label": "Custom Submenu 3"
            }
        ],
        "menus": {
            "cjg.customsubmenu1": [
                {
                    "command": "workbench.view.explorer"
                }
            ],
            "ep.custosubmenu1": [
                {
                    "command": "view.zoomOut"
                },
                {
                    "submenu": "ep.custosubmenu2"
                },
                {
                    "submenu": "cjg.customsubmenu1" // Direct child of submenu 1 won't appear because it will be blocked by child of submenu 2
                }
            ],
            "ep.custosubmenu2": [
                {
                    "command": "view.zoomIn",
                    "group": "0_ep"
                },
                {
                    "submenu": "ep.custosubmenu1"
                },
                {
                    "submenu": "cjg.customsubmenu1" // Will appear under submenu 2 because it will be processed first by tree traversal
                }
            ],
            "explorer/context": [
                {
                    "command": "reset.layout"
                },
                {
                    "submenu": "ep.custosubmenu1"
                }
            ],
            "editor/title": [
                {
                    "submenu": "ep.custosubmenu1"
                },
                {
                    "submenu": "ep.custosubmenu1"
                }
            ]
        }
}
```

and that led to the following submenu (note that `Custom Submenu 3` only appears once under `custom submenu 2` because that submenu was processed by your tree traversal before the `Custom Submenu 3` that would have been a direct child of `custom submenu 1`)

![image](https://user-images.githubusercontent.com/62660806/116138510-861aa680-a69a-11eb-942d-7fc6c10931b0.png)

Reply:
You are absolutely right! I made changes according to what you wrote but still haven't pushed them... 
So the correct behavior you should test after pushing, Sorry for that.

Reply:
Suppose to work now